src/memory.js

```diff
@@ -1,18 +1,39 @@
 class MemoryGame {
   constructor(cards) {
     this.cards = cards;
-    // add the rest of the class properties here
+    this.pickedCards = [];
+    this.pairsClicked = 0;
+    this.pairsGuessed = 0;
   }
 
   shuffleCards() {
-    // ... write your code here
+    if (!this.cards) {
+      return undefined;
+
+    } else {
+      return this.cards.sort(() => Math.random() - 0.5);
+
+    }
   }
 
   checkIfPair(card1, card2) {
-    // ... write your code here
+    this.pairsClicked++;
+  
+    if (card1 === card2) {
+      this.pairsGuessed++;
+      return true;
+
+    } else {
+      return false;
+
+    }
   }
 
-  checkIfFinished() {
-    // ... write your code here
+  checkIfFinished(totalPairs) {
+    if (this.pairsGuessed === this.totalPairs) {
+      return true;
+    } else {
+      return false;
+    }
   }
```

Comment:
This.totalPairs no lo tienes declarado dentro de la clase del Memory, por eso a ese punto ese elemento es undefined, ahora bien al metodo checkIfFinished no le llega ningun argumento, tienes que sacar ese totalPairs de otro sitio, en el fondo sabes donde se encuentra, ya que si tienes el numero total de cards, para saber cuantas parejas hay, lo unico que tenemos que hacer es dividir el total de cards entre 2, es decir:

```suggestion
  checkIfFinished() {
    if (this.pairsGuessed === this.cards.length / 2) {
      return true;
    } else {
      return false;
    }
  }
```

Ahora bien este codigo tambien se puede modificar de la sigueinte forma, ya que la unica respuesta de aqui es true o false puedes devolver lo que resuelve la pregunta del if:

```suggestion
  checkIfFinished() {
    return this.pairsGuessed === this.cards.length / 2
  }
```